### PR TITLE
library/browse/browsetablemodel: Fix QList::insert warning

### PR DESCRIPTION
**Pull request by Holzhaus**

### CODE REVIEW

**Comment by Be-ing** [comment]
PRs for code quality issues should go to the master branch unless there is a bug which has a real world impact on users.

**Comment by Holzhaus** [comment]
> PRs for code quality issues should go to the master branch unless there is a bug which has a real world impact on users.

I disagree. I'm okay with merging it into 2.3 instead of 2.2, but these kinds of issues are bugs and need to be fixed. In this case It no only prints a warning to the log on each run, it also makes debugging harder because QT_FATAL_WARNINGS will make mixxx crash due to this issue and I can't debug problems that pop up later on.

**Comment by Be-ing** [comment]
Okay, we can target this at the beta branch. Changes should *only* be made to stable branches for bugs that have a practical impact on users. We have had regressions in stable releases before when we haven't been strict about that.

**Comment by Holzhaus** [comment]
> What issue does this PR fix? I don't see any warning.

    Warning [Main]: QList::insert(): Index out of range.

Right after the ALSA/JACK init messages.

**Comment by uklotzde** [comment]
> > What issue does this PR fix? I don't see any warning.
> 
> ```
> Warning [Main]: QList::insert(): Index out of range.
> ```
> 
> Right after the ALSA/JACK init messages.

No, I am not able to reproduce it.

**Comment by uklotzde** [comment]
Ok, but I understand why. This code is evil, completely broken by design.

**Comment by Holzhaus** [comment]
Yes, I agree.

**Comment by uklotzde** [comment]
Please add a comment before the first insert() that the ordering must exactly match the ordering of the enum that is used for indexing.

**Comment by daschuer** [comment]
You may consider to add a small helper for safty:
```
insertAtList(QList* pList, int index, const T& value) {
    if (pList.size() > index ) {
        pList.replace(index, value);
        return; 
    }
    while (pList.size() < index) {
           pList.append(T()); 
    }
    pList.append(value); 
    DEBUG_ASSERT(pList.size() == index + 1); 
}
```

File: src/util/compatibility.h [hunk 1]
```diff
@@ -217,3 +218,24 @@ inline uint qHash(const QList<T>& key, uint seed = 0) {
...
 }
 #endif
+
+/// Helper to insert values into a QList with specific indices.
+///
+/// *For legacy code only - Do not use for new code!*
+template <typename T>
+inline void listAppendOrReplaceAt(QList<T>* pList, int index, const T& value) {
+    int initialListSize = pList->size();
+    VERIFY_OR_DEBUG_ASSERT(initialListSize >= index) {
```
Comment (uklotzde) [inline src/util/compatibility.h:228]: Why so complicated? More consistent with less noise:

```
VERIFY_OR_DEBUG_ASSERT(index <= pList->size()) {
    qWarning() << "listAppendOrReplaceAt: Padding list with"
                   << (index - pList->size()) << "default elements";
     while (index > pList->size()) {
         pList->append(T());
     }
}
```
DEBUG_ASSERT(index <= pList->size());
if (index < pList->size()) {
    pList->replace(index, value);
} else {
    pList->append(value);
}

**Review by daschuer — APPROVED** [review]
Thank you, LGTM. 

We have some more places where QList::insert() is used in this unsaved way. 
It would be nice if we can fix that as well. This is no reason for not merging this though.

**Review by uklotzde — APPROVED** [review]

**Pull request merged**